pyomo/contrib/pynumero/interfaces/pyomo_nlp.py

```diff
@@ -1018,3 +1117,283 @@ def load_state_into_pyomo(self, bound_multipliers=None):
+                self._additional_output_entries_data = -1.0*np.ones(self._ex_model.n_outputs())
+
+            col = self._outputs_jac_primal_jcol
+            data = outputs_jac.data
+
+            # add the additional entries for the -Po*x portion of the jacobian
+            row = np.concatenate((row, self._additional_output_entries_irow))
+            col = np.concatenate((col, self._additional_output_entries_jcol))
+            data  = np.concatenate((data, self._additional_output_entries_data))
+            outputs_jac = coo_matrix(
+                (data, (row, col)),
+                shape=(outputs_jac.shape[0], self._n_primals))
+
+        jac = None
+        if eq_jac is not None:
+            if outputs_jac is not None:
+                # create a jacobian with both Jw_eq and Jw_o
+                jac = BlockMatrix(2,1)
+                jac.name = 'external model jacobian'
+                jac.set_block(0,0,eq_jac)
+                jac.set_block(1,0,outputs_jac)
+            else:
+                assert self._ex_model.n_outputs() == 0
+                assert self._ex_model.n_equality_constraints() > 0
+                # only need the Jacobian with Jw_eq (there are not
+                # output equations)
+                jac = eq_jac
+        else:
+            assert outputs_jac is not None
+            assert self._ex_model.n_outputs() > 0
+            assert self._ex_model.n_equality_constraints() == 0
+            # only need the Jacobian with Jw_o (there are no equalities)
+            jac = outputs_jac
+
+        return jac
+
+    def evaluate_hessian(self):
+        # compute the portion of the Hessian of the Lagrangian
+        # of h(x) w.r.t. x
+        # H_h(x) = sum_i y_eq_i * Hw_eq(Pu*x) +_ sum_k y_o_j * Hw_o(Pu*x)]
+
+        data_list = []
+        irow_list = []
+        jcol_list = []
+
+        # Hw_eq(x)
+        eq_hess = None
+        if self._ex_model.n_equality_constraints() > 0:
+            eq_hess = self._ex_model.evaluate_hessian_equality_constraints()
+            if self._eq_hess_jcol is None:
+                # first time through, let's also check that it is lower triangular
+                if np.any(eq_hess.row - eq_hess.col < 0):
+                    raise ValueError('ExternalGreyBoxModel must return lower '
+                                     'triangular portion of the Hessian only')
+
+                # The first time through, we won't have created the
+                # mapping of external primals ('u') to the full space
+                # primals ('x')
+                self._eq_hess_irow = row = self._inputs_to_primals_map[eq_hess.row]
+                self._eq_hess_jcol = col = self._inputs_to_primals_map[eq_hess.col]
+
+                # mapping may have made this not lower triangular
+                mask = col > row
+                if np.any(mask):
+                    temp = row[mask].copy()
+                    row[mask] = col[mask]
+                    col[mask] = temp
```

Comment:
The following python idiom avoids a copy (and the 'if' is unnecessary):
```
mask = col > row
row[mask], col[mask] = col[mask], row[mask]
```

Reply:
yep. changed.